Tribler/Core/TorrentChecker/session.py

```diff
@@ -198,309 +185,384 @@ def __init__(self, tracker_url, tracker_address, announce_page, on_result_callba
-            self._content_length = int(self._header_buffer[idx:].split('\r\n', 1)[0].strip())
-
-    def _process_scrape_response(self):
-        # parse the retrieved results
-        if self._message_buffer is None:
-            return False
-        response_dict = bdecode(self._message_buffer)
-        if response_dict is None:
-            return False
+        seed_leech_dict = {}
 
         unprocessed_infohash_list = self._infohash_list[:]
         if 'files' in response_dict and isinstance(response_dict['files'], dict):
             for infohash in response_dict['files']:
-                downloaded = response_dict['files'][infohash].get('downloaded', 0)
                 complete = response_dict['files'][infohash].get('complete', 0)
+                downloaded = response_dict['files'][infohash].get('downloaded', 0)
                 incomplete = response_dict['files'][infohash].get('incomplete', 0)
 
-                seeders = downloaded
+                # Sow complete as seeders. "complete: number of peers with the entire file, i.e. seeders (integer)"
+                #  - https://wiki.theory.org/BitTorrentSpecification#Tracker_.27scrape.27_Convention
+                seeders = complete
                 leechers = incomplete
 
-                # handle the retrieved information
-                self._on_result_callback(infohash, seeders, leechers)
+                # Store the information in the dictionary
+                seed_leech_dict[infohash] = (seeders, leechers)
 
                 # remove this infohash in the infohash list of this session
                 if infohash in unprocessed_infohash_list:
                     unprocessed_infohash_list.remove(infohash)
 
         elif 'failure reason' in response_dict:
-            self._logger.debug(u"%s Failure as reported by tracker [%s]", self, repr(response_dict['failure reason']))
-
-            return False
+            self._logger.info(u"%s Failure as reported by tracker [%s]", self, repr(response_dict['failure reason']))
+            self.failed()
+            return
 
         # handle the infohashes with no result (seeders/leechers = 0/0)
         for infohash in unprocessed_infohash_list:
             seeders, leechers = 0, 0
-            # handle the retrieved information
-            self._on_result_callback(infohash, seeders, leechers)
-        return True
+            seed_leech_dict[infohash] = (seeders, leechers)
 
+        self._is_finished = True
+        self.result_deferred.callback(seed_leech_dict)
 
-class UdpTrackerSession(TrackerSession):
+    def cleanup(self):
+        """
+        Cleans the session by cancelling all deferreds and closing sockets.
+        :return: A deferred that fires once the cleanup is done.
+        """
+        super(HttpTrackerSession, self).cleanup()
+        if self.request is not None:
+            self.request.cancel()
+
+        if self.result_deferred:
+            self.result_deferred.cancel()
```

Comment:
Why can't it be cancelled? If you shutdown tribler all session present [will be cleaned using the `cleanUp` function](https://github.com/Tribler/tribler/pull/2038/files#diff-9c4b6475dbed144b806a02887d2e1cc3R79). It makes no sense invoking either the callback or errback in this case as both of them will then invoke code to [update the state of the tracker](https://github.com/Tribler/tribler/pull/2038/files#diff-9c4b6475dbed144b806a02887d2e1cc3R271). You want nothing to happen as you are in an intermediate state where you have to shutdown. Also, the cancel() function will do nothing if the Deferred has already fired either it's errback or callback.